src/main/scala/sparkz/core/network/peer/InMemoryPeerDatabase.scala

```diff
@@ -28,52 +28,33 @@ final class InMemoryPeerDatabase(settings: NetworkSettings, timeProvider: TimePr
     */
   private var penaltyBook = Map.empty[InetAddress, (Int, Long)]
 
-  override def get(peer: InetSocketAddress): Option[PeerInfo] = peers.get(peer).map(_._1)
-
-  private def addPeer(peerInfo: PeerInfo): Unit = {
-    peerInfo.peerSpec.address.foreach { address =>
-      log.debug(s"Updating peer info for $address")
-      peers += address -> (peerInfo, timeProvider.time())
+  override def get(peer: InetSocketAddress): Option[PeerInfo] =
+    bucketManager.getPeer(peer) match {
+      case Some(peerBucketValue) => Some(peerBucketValue.peerInfo)
+      case _ => None
     }
-  }
 
   /**
    * Adds a peer to the in-memory database ignoring the configurable limit.
    * Used for high-priority peers, like peers from config file or connected peers
    */
-  override def addOrUpdateKnownPeer(peerInfo: PeerInfo): Unit = {
+  override def addOrUpdateKnownPeer(peerInfo: PeerInfo, source: Option[ConnectedPeer]): Unit = {
```

Comment:
Here we're not considering when we receive an incoming connection. If the connection flow is successful, then we try to insert into the database the new peer's address, but we don't know anything about its source (since it's an incoming connection and nobody shared it), so it wrongly treats the connection like it's already in the new db and we have to move it to the tried, causing the exception to be thrown. We need to add the source when the [handshake process ends](https://github.com/HorizenOfficial/Sparkz/blob/main/src/main/scala/sparkz/core/network/NetworkController.scala#L360).